### PR TITLE
Merge Lottery Gift Box (Continued Receiving)

### LINKED ISSUE #757: Option to Move to Gift Box Script from the Lottery Script, And Stop Only When Inventory is Full
Before opening an issue:
- [x] No one has asked for this feature before in other issues (even the closed ones)

> If you open unnecessary/dumb issues, you'll be rewarded with appropriate tags.

Since TW and JP both have lottery events I think some people may have some enhancement ideas, so I would like to pitch in some of my ideas when I was opening my lottery boxes:

- #476 Mentioned combining Lottery/Gift Box scripts together. While combining those scripts entirely is not practical, I find myself wanting the Lottery script to click on the "move to gift box" button (and even want it to just start clicking). Judging from the Jetpack Compose UI, could we add a toggle to say if we want to the gift box script to run right after lottery, and link those two together?
- Gift Box script now selects up to 99 exp stacks and exits. Is there a way we can make it click on the receive button then pick it again, and only stop when FGO ask us to sell exps or move to servant enhancement (or even click on the burn button to go to the shop)?

Basically,
- adding an option to click on the "move to gift box" button and initiate gift box script, and
- making the gift box script receive the selected exp stacks until inventory is full
should make our box opening experience even better.

Let me know if there's any technical difficulties and how I can help/test it, thanks!

### PR DESCRIPTION
**Pull request by the3dsandwich**
Fixes #757. Except I don't know how to use screen and map so connection failure is written using a dumb while loop.

https://user-images.githubusercontent.com/32169552/124288512-d7a44d80-db83-11eb-9595-4ffcd0e4d7f4.mp4

Video (if upload is successful shows the script continues to receive picked embers until inventory is full (debugged using a lower max click count)

### CODE REVIEW

**Review by the3dsandwich — COMMENTED** [review]
Added delay to give FGO enough time to switch to gift box, so gift box script won't give the "Couldn't find Embers on screen. This shouldn't happen." exception.

**Comment by MathewSachin** [comment]
The video isn't loading lol. Can you upload it to YouTube (as an unlisted video)?

**Comment by the3dsandwich** [comment]
Got another video, taken at bbd5d23. After that I added the 1 second delay. Long video alert

https://youtu.be/XPNuUlvTgIQ

File: scripts/src/main/java/com/mathewsachin/fategrandautomata/scripts/entrypoints/AutoGiftBox.kt [hunk 1]
```diff
@@ -31,18 +33,24 @@ class AutoGiftBox @Inject constructor(
         val checkRegion = Region(xpOffsetX + 1320, 350, 140, 1500)
         val scrollEndRegion = Region(100 + checkRegion.X, 1421, 320, 19)
         val receiveSelectedClick = Location(1890 + xpOffsetX, 750)
+        val receiveEnabledRegion = Region(2000, 400, 300, 100)
```
Comment (MathewSachin) [inline scripts/src/main/java/com/mathewsachin/fategrandautomata/scripts/entrypoints/AutoGiftBox.kt:36]:
```suggestion
        val receiveEnabledRegion = Region(1755 + xpOffsetX, 410, 290, 60)
```

Adjusting for JP. Can you check if these coordinates still work?

**Comment by MathewSachin** [comment]
That video says it's private which means only you can watch it. Change it to unlisted.

I made a change for JP support. Can you check if that still works on TW?

**Comment by the3dsandwich** [comment]
> That video says it's private which means only you can watch it. Change it to unlisted.

Was uploading. Now it's saved to unlisted.

> I made a change for JP support. Can you check if that still works on TW?

Testing!

**Comment by the3dsandwich** [comment]
> I made a change for JP support. Can you check if that still works on TW?

Yup, works just fine!

**Comment by MathewSachin** [comment]
Saw the video. It's working great!
So, the only problem we have is that if embers aren't received successfully in that 2s time frame, the script will stop.
This is a pretty good trade-off I suppose.

File: scripts/src/main/java/com/mathewsachin/fategrandautomata/scripts/entrypoints/AutoGiftBox.kt [hunk 2]
```diff
@@ -31,18 +33,24 @@ class AutoGiftBox @Inject constructor(
...
         while (true) {
+            val receiveEnabledPattern = receiveEnabledRegion.getPattern()
             val picked = iteration(checkRegion, scrollEndRegion)
             totalReceived += picked
 
             if (picked > 0) {
                 receiveSelectedClick.click()
+                while (true) {
+                    Duration.seconds(2).wait()
+                    if(needsToRetry()) retry() else break
```
Comment (MathewSachin) [inline scripts/src/main/java/com/mathewsachin/fategrandautomata/scripts/entrypoints/AutoGiftBox.kt:47]:
```suggestion
                    if (needsToRetry()) retry() else break
```

just a nitpick

**Comment by the3dsandwich** [comment]
I suppose there can be more optimizations to the scripts, but for now I think it's good enough. Thanks!

**Pull request merged**